### PR TITLE
Improving Starlight laser handling

### PR DESCRIPTION
**Pull request by danfireman**
Fixes https://github.com/ZeroK-RTS/Zero-K/issues/3729 

TODO:
- [x] Tidy
- [x] ~~Stop ignoring Avoid Bad Targets~~ wontfix, not really much need for this toggle now that targeting has been improved
- [x] Activate for Bertha and DRP

### CODE REVIEW

File: LuaUI/Widgets/unit_starlight_targetting.lua [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+				self.targetPos = {targetX, targetY, targetZ}
+			end
+			self.currentTarget = unitIDorPos
+		end
+	end
+}
+StarlightControllerMT = {__index = StarlightController}
+
+function widget:UnitFinished(unitID, unitDefID, unitTeam)
+		if (unitDefID == Starlight_ID)
```
Comment (sprunk) [inline LuaUI/Widgets/unit_starlight_targetting.lua:123]: see line 31-32
```suggestion
		if (unitDefID == StarlightUnitDefID)
```

File: LuaUI/Widgets/unit_starlight_targetting.lua [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+local GetUnitDefID = Spring.GetUnitDefID
+local GetTeamUnits = Spring.GetTeamUnits
+local GetUnitStates = Spring.GetUnitStates
+local Echo = Spring.Echo
+local GetSpecState = Spring.GetSpectatingState
+local CMD_STOP = CMD.STOP
+local CMD_ATTACK = CMD.ATTACK
+local CMD_UNIT_SET_TARGET = 34923
+
+local Starlight_ID = UnitDefNames.mahlazer.id
```
Comment (sprunk) [inline LuaUI/Widgets/unit_starlight_targetting.lua:31]: Same as the next line, drop
```suggestion
```

File: LuaUI/Widgets/unit_starlight_targetting.lua [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+function sqdist(p, q)
+	return sq(p[1]-q[1]) + sq(p[2]-q[2]) + sq(p[3]-q[3])
+end
+
+local StarlightControllerMT
+local StarlightController = {
+	unitID,
+	pos,
+	targetPos,
+	currentTarget,
```
Comment (sprunk) [inline LuaUI/Widgets/unit_starlight_targetting.lua:54]: comment out (atm it's referencing some random nil globals)
```suggestion
	-- unitID, pos, targetPos, currentTarget
```

**Review by sprunk — CHANGES_REQUESTED** [review]
minor technicalities

**Comment by GoogleFrog** [comment]
I dislike metatables. Are functions being copied multiple times here? I don't see what would be wrong with declaring a bunch of local functions and passing simple tables of local data to them. The handling part of the widget could even be cleaned up with IterableMap.

File: LuaUI/Widgets/unit_starlight_targetting.lua [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+
+	getTargetToClosest = function(self)
+		if self.targetPos ~= nil then
+			local nearUnits = Spring.GetUnitsInRectangle(self.targetPos[1]-2000, self.targetPos[3]-2000, self.targetPos[1]+2000, self.targetPos[3]+2000)
+			local shortestDist = 9999999
+			local bestSol = nil
+			for k, v in pairs(nearUnits) do
+				if not (Spring.IsUnitAllied(v)) and (Spring.IsUnitInLos(v) or immobiles[Spring.GetUnitDefID(v)]) then
+					local x,y,z = Spring.GetUnitPosition(v)
+					local dist = sqdist(self.targetPos, {x,y,z})
```
[the remark this answers is not in the document]
Comment (GoogleFrog) [reply LuaUI/Widgets/unit_starlight_targetting.lua:82]: This also spams a lot of tables.

File: LuaUI/Widgets/unit_starlight_targetting.lua [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+		self.unitID = unitID
+		self.pos = {GetUnitPosition(self.unitID)}
+		self.targetPos = nil
+		self.currentTarget = nil
+		return self
+	end,
+
+	unset = function(self)
+		--Echo("StarlightController removed:" .. self.unitID)
+		GiveOrderToUnit(self.unitID,CMD_STOP, {}, {""},1)
```
[the remark this answers is not in the document]
Comment (danfireman) [reply LuaUI/Widgets/unit_starlight_targetting.lua:65]: I copy-pasted this from somewhere. Shall change to 0

File: LuaUI/Widgets/unit_starlight_targetting.lua (hunk 4 above)
[the remark this answers is not in the document]
Comment (danfireman) [reply LuaUI/Widgets/unit_starlight_targetting.lua:77]: Not too sure I follow. Are you disliking the `{x,y,z}`?

File: LuaUI/Widgets/unit_starlight_targetting.lua (hunk 5 above)
[the remark this answers is not in the document]
Comment (danfireman) [reply LuaUI/Widgets/unit_starlight_targetting.lua]: Done

**Comment by danfireman** [comment]
> I dislike metatables. Are functions being copied multiple times here? I don't see what would be wrong with declaring a bunch of local functions and passing simple tables of local data to them. The handling part of the widget could even be cleaned up with IterableMap.

Note that I've changed this now.

**Comment by GoogleFrog** [comment]
Why was bait_level_default removed?

**Comment by GoogleFrog** [comment]
spring_2022-05-22_15-04-59
The widget is now fine technically, but I'm not happy with how poorly it works with latency. I'd at least want an easy way to disable it (perhaps by checking anti-bait state?).

**Pull request closed without merging**